### PR TITLE
Allow TopicLeaderDistributionGoal to swap follower replicas

### PR DESCRIPTION
**Pull request by andrey-klochkov-liftoff**
In some cases a broker than leads too many partitions can’t pass leadership to any of its follower brokers because each follower leads max amount of partitions already, and brokers that don’t lead enough partitions are not among the followers. This change allows `TopicLeaderDistributionGoal` to “swap” follower replicas of the same topic between brokers when otherwise it’s not able to pass the leadership.

I tested it by deploying this to the `us-e` cruise control, doing a dry run and verifying overall proposal stats and then checking rebalance actions for one of the topics in trace logs of CC analyzer.

### CODE REVIEW

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java [hunk 1]
```diff
@@ -439,63 +439,95 @@ private boolean rebalanceByMovingLeadershipOut(Broker broker,
     if (!clusterModel.deadBrokers().isEmpty()) {
```
Comment (PleasingFungus) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java:459]: Can we add a comment to this function, clarifying that the return value is `true` if and only if we fail to rebalance? The meaning of the return value is currently very obscure.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java [hunk 2]
```diff
@@ -439,63 +439,95 @@ private boolean rebalanceByMovingLeadershipOut(Broker broker,
...
+    TopicPartition tp = leader.topicPartition();
+    String topic = tp.topic();
+    Set<Broker> nonFollowers = clusterModel.brokers().stream()
+            .filter(x -> x != broker && !followers.contains(x))
+            .collect(Collectors.toSet());
+    for (Broker follower : followers) {
+      Optional<Replica> replica = follower.replicasOfTopicInBroker(topic).stream()
+              .filter(r -> r.topicPartition().equals(tp) && !r.isCurrentOffline())
+              .findAny();
+      if (replica.isPresent()) {
```
Comment (PleasingFungus) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java]: Let's reduce indentation here.
```
if (!replica.isPresent()) {
    continue;
}
```

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java [hunk 3]
```diff
@@ -439,63 +439,95 @@ private boolean rebalanceByMovingLeadershipOut(Broker broker,
...
+   * Find a broker that can swap a follower replica with a follower replica of a broker following a given leader replica.
+   * Making such swap allows leadership to be moved to a broker that is not in the list of the immediate followers of
+   * the broker that's leading the partition currently.
+   *
+   * @param clusterModel        The state of the cluster.
+   * @param leader              Leader replica to be applied the given balancing action.
+   * @param followers           Brokers currently hosting follower replicas of the leader replica.
+   * @param optimizedGoals      Optimized goals.
+   * @param optimizationOptions Options to take into account during optimization -- e.g. excluded brokers for leadership.
+   * @param callback            A callback that does additional verification that a candidate broker is suitable.
```
Comment (PleasingFungus) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java]: `callback` is a very, very nonspecific name for a variable - it's like calling a function `helper`. Let's call this `brokerIsSuitable` or `checkBrokerSuitability` or something along those lines.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java [hunk 4]
```diff
@@ -439,63 +439,95 @@ private boolean rebalanceByMovingLeadershipOut(Broker broker,
...
-              candidateBrokers,
-              ActionType.LEADERSHIP_MOVEMENT,
-              optimizedGoals,
-              optimizationOptions);
-      // Only check if we successfully moved something.
-      if (b != null) {
-        if (++numLeaderReplicas >= _balanceLowerLimitByTopic.get(topic)) {
-          return false;
+  /**
+   * Find a broker that can swap a follower replica with a follower replica of a broker following a given leader replica.
```
Comment (PleasingFungus) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java]: This sentence is pretty hard to follow. It's also misleading - it suggests we're looking for a broker that *can* swap a follower, but this function also actually *performs* that swap. Currently, the function doc only mentions that much further down, in the return value.

So instead of this first sentence, what do you think about the rather longer:
```
   * Attempt to swap one broker's follower replica of a given topic-partition away. This swap will only be
   * performed in exchange for another follower replica of the same topic (but a different partition), and
   * will not allow any broker to be left with multiple replicas of the same topic-partition.
```

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java [hunk 5]
```diff
@@ -439,63 +439,95 @@ private boolean rebalanceByMovingLeadershipOut(Broker broker,
...
+   * Making such swap allows leadership to be moved to a broker that is not in the list of the immediate followers of
+   * the broker that's leading the partition currently.
```
Comment (PleasingFungus) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java]: This second sentence would make more sense at the place this function is called (currently line 460). It describes one way that this function can be used, but not what it does. One can imagine this function being used for some other purpose later, at which point this comment would be inaccurate/misleading. 

Also, grammar nit: This should either be `Making such a swap` or `Making this swap`.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java [hunk 6]
```diff
@@ -439,63 +439,95 @@ private boolean rebalanceByMovingLeadershipOut(Broker broker,
...
+                .map(Replica::topicPartition)
+                .collect(Collectors.toSet());
+        for (Broker nonFollower : nonFollowers) {
+          Set<Replica> candidateReplicas = nonFollower.replicasOfTopicInBroker(topic).stream()
+                  .filter(r -> !r.isLeader() && !r.isCurrentOffline())
+                  .filter(r -> !followerPartitions.contains(r.topicPartition()))
+                  .collect(Collectors.toSet());
+          Replica r = maybeApplySwapAction(clusterModel, replica.get(), new TreeSet<>(candidateReplicas), optimizedGoals, optimizationOptions);
+          if (r != null) {
+            if (callback.process(nonFollower)) {
```
Comment (PleasingFungus) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java]: Does this need to be a callback? Can't we just embed the logic inline?

```
          if (maybeApplyBalancingAction(clusterModel, leader, Collections.singleton(nonFollower),
                      ActionType.LEADERSHIP_MOVEMENT,
                      optimizedGoals,
                      optimizationOptions)) {
            return nonFollower;
          }
```

Seems easier to follow to me, but I haven't tested it and might be overlooking something. WDYT?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java [hunk 7]
```diff
@@ -439,63 +439,95 @@ private boolean rebalanceByMovingLeadershipOut(Broker broker,
...
+                .filter(b -> b != broker && !b.replica(leader.topicPartition()).isCurrentOffline())
+                .collect(Collectors.toSet());
+        Broker b = maybeApplyBalancingAction(clusterModel,
+                leader,
+                candidateBrokers,
+                ActionType.LEADERSHIP_MOVEMENT,
+                optimizedGoals,
+                optimizationOptions);
+
+        if (b == null && withFollowerSwaps) {
```
Comment (PleasingFungus) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java]: When `withFollowerSwaps` is true, we've already iterated over all of these brokers, so we know that `b` will be null already, right?

The simple change here is something like:
```
Set<Broker> candidateBrokers = clusterModel.partition(...)
if (withFollowerSwaps)
{
    b = maybeSwapFollowerReplica(...)
} else {
   b = maybeApplyBalancingAction(...)
}
if (b != null ...) {
    return false
}
```

I vaguely wonder about removing the outer `withFollowerSwaps` loop entirely and just having two loops, but maybe that duplicates too much code.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java [hunk 8]
```diff
@@ -343,11 +350,10 @@ private static boolean skipBrokerRebalance(Broker broker,
                                              ClusterModel clusterModel,
                                              Collection<Replica> replicas,
                                              boolean requireLessReplicas,
-                                             boolean requireMoreReplicas,
                                              boolean hasOfflineTopicReplicas,
                                              boolean moveImmigrantReplicaOnly) {
```
Comment (PleasingFungus) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java]: Hm, what's the purpose of the changes to this function?

**Review by PleasingFungus — CHANGES_REQUESTED** [review]

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java (hunk 8 above)
Comment (andrey-klochkov-liftoff) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java]: Previously the goal did two kinds of rebalances: for brokers that led too many partitions it moved the leadership "out", and for brokers that led too few it moved the leadership "in". I thought that we actually don't need both, it's enough to fix the balance for brokers that have lead too many partitions, and that would automatically fix the balance for brokers that have too few.

Actually, now I realized that when the ideal balance is not achievable, not doing two-way balancing actually may leave some brokers leading fewer partitions and they would violate the lower boundary condition, and in general the overall distribution of the load between brokers may be less worse than it could be. I'll revert this part of the change and will do two-way rebalancing, in particular doing replica swaps in both cases two.

Comment (PleasingFungus) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java]: Thanks, that sounds reasonable.

**Comment by andrey-klochkov-liftoff** [comment]
@PleasingFungus , I realized that I can do a single swap operation instead of a swap followed by leadership movement and so I simplified the implementation, making most of your comments non relevant. I'm sorry I figured this out just now and that you had to review a very different version.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java (hunk 7 above)
Comment (andrey-klochkov-liftoff) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java]: I removed this loop altogether.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java (hunk 6 above)
Comment (andrey-klochkov-liftoff) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java]: This is not applicable anymore as I switched to a simpler approach that doesn't use a separate method for finding replicas to swap.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java (hunk 5 above)
Comment (andrey-klochkov-liftoff) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java]: This is not applicable anymore as I switched to a simpler approach that doesn't use a separate method for finding replicas to swap and so I deleted this method.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java (hunk 4 above)
Comment (andrey-klochkov-liftoff) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java]: This is not applicable anymore as I switched to a simpler approach that doesn't use a separate method for finding replicas to swap.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java (hunk 2 above)
Comment (andrey-klochkov-liftoff) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java]: This is not applicable anymore as I switched to a simpler approach and this code doesn't exist anymore.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java [hunk 9]
```diff
@@ -431,6 +439,18 @@ protected void rebalanceForBroker(Broker broker,
...
+   * Attempt to decrease the number of topic partitions led by the broker first by moving the leadership to follower
+   * replicas, and second by swapping leader replicas with follower replicas of partitions not hosted by the broker
+   * currently.
+   *
+   * @param broker              Broker that leads too many partitions of the given topic.
+   * @param topic               Topic to rebalance.
+   * @param clusterModel        The state of the cluster.
+   * @param optimizedGoals      Optimized goals.
+   * @param optimizationOptions Options to take into account during optimization.
+   * @return true if rebalancing was not successful.
```
Comment (PleasingFungus) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java:452]: Did you consider inverting the return value here? It seems easier to reason about if this `returns true if the rebalancing was successful`. (This suggestion is optional.)

Likewise for `rebalanceByMovingLeadershipOut`.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java [hunk 10]
```diff
@@ -455,16 +471,45 @@ private boolean rebalanceByMovingLeadershipOut(Broker broker,
...
+      }
+    }
+
+    // Try to rebalance by swapping one of the broker follower replicas with a leader replica from another broker.
+    Set<TopicPartition> brokerPartitions = broker.replicasOfTopicInBroker(topic).stream()
+            .map(Replica::topicPartition)
+            .collect(Collectors.toSet());;
+    for (Replica leader : broker.leadersOfTopicInBroker(topic)) {
+      Iterable<Broker> candidateBrokers = clusterModel.brokers().stream()
+              .filter(b -> broker.numLeadersOfTopicInBroker(topic) <= _balanceUpperLimitByTopic.get(topic))
```
Comment (PleasingFungus) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java]: Hm, shouldn't this be `< _balanceUpperLimitByTopic.get(topic)`? If the other broker is already maxed out on leadership for the given topic (`== balanceUpperLimit`), then we won't be able to transfer leadership of the partition to it, right?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java [hunk 11]
```diff
@@ -455,16 +471,45 @@ private boolean rebalanceByMovingLeadershipOut(Broker broker,
...
+              .collect(Collectors.toList());
+      for (Broker candidateBroker : candidateBrokers) {
+        List<Replica> candidateReplicas = candidateBroker.replicasOfTopicInBroker(topic).stream()
+                .filter(r -> !r.isLeader() && !r.isCurrentOffline() && !brokerPartitions.contains(r.topicPartition()))
+                .collect(Collectors.toList());
+        Replica r = maybeApplySwapAction(clusterModel, leader, new TreeSet<>(candidateReplicas), optimizedGoals, optimizationOptions);
+        if (r != null && broker.numLeadersOfTopicInBroker(topic) >= _balanceLowerLimitByTopic.get(topic)) {
```
Comment (PleasingFungus) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java]: > `broker.numLeadersOfTopicInBroker(topic) >= _balanceLowerLimitByTopic.get(topic)`

If we're concerned that moving one leader away from the current broker might lead to it dropping below the lower limit for topic leadership, shouldn't we just check that `broker.numLeadersOfTopicInBroker(topic) > _balanceLowerLimitByTopic.get(topic)` at the very start of this function? I'm very confused about what this check is doing here, or what we expect to happen if `r != null` but the original broker is now below the topic leadership limit - we'll keep trying to swap this replica's leadership around between every possible broker?

It looks like this might be a copy-paste error. Were you actually trying to double-check that `broker.numLeadersOfTopicInBroker(topic) < _balanceUpperLimitByTopic.get(topic)`?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java [hunk 12]
```diff
@@ -488,13 +532,34 @@ private boolean rebalanceByMovingLeadershipIn(Broker broker,
...
+        return false;
+      }
+    }
+
+    // Try to rebalance by swapping one of the broker non-leader replicas with a leader replica from another broker.
+    Set<TopicPartition> brokerPartitions = broker.replicasOfTopicInBroker(topic).stream()
+            .map(Replica::topicPartition)
+            .collect(Collectors.toSet());
+    candidateBrokers = clusterModel.brokers().stream()
+            .filter(b -> broker.numLeadersOfTopicInBroker(topic) >= _balanceLowerLimitByTopic.get(topic))
```
Comment (PleasingFungus) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java]: Seems like this should check that `broker.numLeadersOfTopicInBroker(topic) > _balanceLowerLimitByTopic.get(topic)`. Otherwise, if we move a partition away from that broker, we'll drop it below the limit, right?

**Review by PleasingFungus — APPROVED** [review]
Looks good, assuming we agree about the few changes I'm suggesting here.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java (hunk 9 above)
Comment (andrey-klochkov-liftoff) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java:452]: This code was initially copied from other CC goals. The implementation of these functions is different (as we're doing per-topic distribution) but the interface is almost exactly the same. It'd be very confusing if the return values of functions with the same name and almost exactly the same interface would have the opposite meaning.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java (hunk 10 above)
Comment (andrey-klochkov-liftoff) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java]: Good point. BTW this condition here is actually a nice to have optimization. For every proposed action CC would check if it satisfies the constraints and reject the bad ones. I added this condition just to skip going through proposals that would be rejected anyway. In the rest of the CC code they don't bother too much filtering out proposals that are known to be rejected, I guess for the sake of code simplicity.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java (hunk 9 above)
Comment (PleasingFungus) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java:452]: OK, fair enough.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java (hunk 11 above)
Comment (andrey-klochkov-liftoff) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java]: > If we're concerned that moving one leader away from the current broker might lead to it dropping below the lower limit for topic leadership, shouldn't we just check that broker.numLeadersOfTopicInBroker(topic) > _balanceLowerLimitByTopic.get(topic) at the very start of this function? I'm very confused about what this check is doing here, or what we expect to happen if r != null but the original broker is now below the topic leadership limit - we'll keep trying to swap this replica's leadership around between every possible broker?

This method is only called for brokers that lead too many partitions. Decreasing the number of leaders on such broker by one can not bring the number of partitions it leads lower than the limit. The condition `r != null` means that the proposed action was successful, i.e. that the number of partitions the broker leads was decreased by 1. I don't see any issues here.

I did have a copy-paste error here, thanks!

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java (hunk 12 above)
Comment (andrey-klochkov-liftoff) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicLeaderDistributionGoal.java]: Correct. It's a nice to have condition anyway, but making it stricter will reduce the number of actions we go through that we know will be rejected.

**Comment by andrey-klochkov-liftoff** [comment]
@PleasingFungus , I fixed a couple of copy-paste errors and made candidate broker conditions stricter as you proposed. Please check my other comments. Are you OK merging it now?

**Comment by PleasingFungus** [comment]
Yes, this looks good now.
